robotsim.py

```diff
@@ -48,118 +48,198 @@
+        pivot = pygame.math.Vector2(self.size//2, -self.size//2)
+        pivot_rotate = pivot.rotate(self.w)
+        pivot_move = pivot_rotate - pivot
+        origin = (self.x - self.size//2 + min_box[0] - pivot_move[0], self.y - self.size//2 - max_box[1] + pivot_move[1])
+        rotated_image = pygame.transform.rotate(robotImg, self.w)
+        gameDisplay.blit(rotated_image, origin)
         pygame.display.update()
         clock.tick(120)
 
     def move_forward(self):
-        angle = self.w
-        x1 = self.x 
-        y1 = self.y 
-        rad = math.radians(angle)
-        x2 = round(math.cos(rad))  + x1
-        y2 = y1 - round(math.sin(rad))
-        generate_map()
-        # Check collisions
-        px1 = self.x + int(pixel_constant * 0.5) + round(math.cos(rad)*self.size*0.5) 
-        py1 = self.y + int(pixel_constant * 0.5) - round(math.sin(rad)*self.size*0.5)
-        for i in range(-int(self.size*0.5)+1,int(self.size*0.5)-1): 
-            px2 = round(math.cos(math.radians(90+angle)) * i) + px1
-            py2 = py1 - round(math.sin(math.radians(90+angle)) * i)
-            is_wall = gameDisplay.get_at((px2,py2)) == colors['black']
-            if is_wall:
-                return
-        self.set_position(x2,y2,angle)
+        if self.ultrasonicFront() > 0 :
+            if self.dir == 0:
+                self.row -= 1
+            if self.dir == 1:
+                self.col -= 1
+            if self.dir == 2:
+                self.row += 1
+            if self.dir == 3:
+                self.col += 1
+            for _ in range(pixel_constant):
+                angle = self.w
+                x1 = self.x 
+                y1 = self.y 
+                rad = math.radians(angle)
+                x2 = round(math.cos(rad)) + x1
+                y2 = y1 - round(math.sin(rad))
+                generate_map()
+                self.set_position(x2,y2,angle)
 
     def move_backward(self):
-        angle = self.w
-        x1 = self.x 
-        y1 = self.y 
-        rad = math.radians(angle)
-        x2 = x1 - round(math.cos(rad))
-        y2 = round(math.sin(rad)) + y1
-        generate_map()
-        # Check collisions
-        px1 = self.x + int(pixel_constant * 0.5) - round(math.cos(rad)*self.size*0.5) 
-        py1 = self.y + int(pixel_constant * 0.5) + round(math.sin(rad)*self.size*0.5)
-        for i in range(-int(self.size*0.5)+1,int(self.size*0.5)-1): 
-            px2 = round(math.cos(math.radians(90+angle)) * i) + px1
-            py2 = py1 - round(math.sin(math.radians(90+angle)) * i)
-            is_wall = gameDisplay.get_at((px2,py2)) == colors['black']
-            if is_wall:
-                return
-        self.set_position(x2,y2,angle)
+        #TODO
+        pass 
     
     def rotate_right(self):
-        generate_map()
-        self.set_position(self.x,self.y,self.w - 1)
+        if 1 <= self.dir <= 3:
+            self.dir -= 1
+        else:
+            self.dir = 3
```

Comment:
You could do something like this (self.dir  - 1 + 4 )%4

Reply:
Just to simplify it and avoid using the if/else conditions

Reply:
Sure!